File: day2/DoublyLinkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }

}

class List {
    constructor() {
        this.head = null;
        this.tail = null;
    }

    addToBack(value) {
        const newNode = new Node(value);
        if (null == this.head) {
            this.head = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
        }
        this.tail = newNode;
    }

    addToFront(value) {
        const newNode = new Node(value);
        if (null == this.head) {
            this.tail = newNode;
        } else {
            this.head.prev = newNode;
            newNode.next = this.head;
        }
        this.head = newNode;

    }

    printForward() {
        let current = this.head;
        while (current != null) {
            console.log(current.value)
            current = current.next
        }
    }

    printBackward() {
        let current = this.tail;
        while (current != null) {
            console.log(current.value)
            current = current.prev
        }
    }

}

ls = new List()
ls.addToBack(3)
ls.addToBack(4)
ls.addToFront(2)
ls.addToFront(1)
console.log('Printing backwards');
ls.printForward()
console.log('Printing backwards');
ls.printBackward()

File: day2/DoublyLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None


class MyDoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def addToBack(self, value):
        node = Node(value)
        if None is self.head:
            self.head = node    
        else:
            self.tail.next = node 
            node.prev = self.tail
        self.tail = node 

    def addToFront(self, value):
        node = Node(value)
        if None is self.head:
            self.tail = node
        else:
            self.head.prev = node 
            node.next = self.head         
        self.head = node 
       
    def printForwards(self):
        current = self.head
        while current is not None:
            print(current.value)
            current = current.next

    def printBackwards(self):
        current = self.tail
        while current is not None:
            print(current.value)
            current = current.prev 

dll = MyDoublyLinkedList()
dll.addToBack(4)
dll.addToBack(5)
dll.addToBack(6)
dll.addToFront(3)
dll.addToFront(2)
dll.addToFront(1)

# dll.printForwards()
dll.printBackwards()